src/Servers/Kestrel/Core/src/Internal/Http/HttpHeaders.cs

```diff
@@ -126,6 +129,18 @@ public void Reset()
             ClearFast();
         }
 
+        protected static string GetInternedHeaderName(string name)
```

Comment:
This method looks like it will be used to set unknown headers.

Aren't most of the headers in this collection known? They will never be used because generated code will handle setting them, right?

Reply:
There are 93 header names in `HeaderNames`; but only 20 in `commonHeaders` shared by `HttpRequestHeaders` and `HttpReponseHeaders`; could drop those 20; or have a different `HashSet<string>` for request and response but it doesn't seem worth it?

However yes, if the header is in the set for `HttpRequestHeaders` or `HttpReponseHeaders` and comes in on that direction it won't use the `HashSet<string>`, it only uses it for header names that aren't in that set rather than making a freshly constructed string long lived.

Any headers outside the set in `HeaderNames` will remain as the newly constructed string and won't be exchanged (i.e. the `HashSet<string>` is not updated while the app is running)

Reply:
Ok!

This two phases of hash lookups (intern hashset then unknown dictionary) makes me wonder whether it makes sense to write a custom collection.

As well as optimizing interning known keys, there is probably a opportunity to optimize memory usage. Most requests won't have that many unknown headers. e.g. an adaptive collection that can switch between `KeyValuePair<string, StringValues>[]` and `Dictionary<string, StringValues>` depending upon how many headers have been added.

Reply:
> Any headers outside the set in `HeaderNames` will remain as the newly constructed string and won't be exchanged 

Is a question on whether to extend the values in `HeaderNames` and it becomes more valuable to do so with this change, rather than just being coinvent for the developer; e.g. missed standardise headers and defacto standard headers: `Prefer`, `Preference-Applied`, `X-Request-ID`, `X-Correlation-ID`, `X-Forwarded-For`, `X-Forwarded-Host`, `X-Forwarded-Proto`, `Refresh`, `X-Powered-By`, `Timing-Allow-Origin`

Though response set headers would likely be less important than request set headers; assuming they are being set by a single assembly so already an interned value.

However that's probably a separate issue...

Reply:
Added issue for other headers https://github.com/dotnet/aspnetcore/issues/31374